rueidiscompat/command.go

```diff
@@ -3452,3 +3462,168 @@ type TSMGetOptions struct {
 	WithLabels     bool
 	SelectedLabels []interface{}
 }
+
+type JSONSetArgs struct {
+	Key   string
+	Path  string
+	Value interface{}
+}
+
+type JSONArrIndexArgs struct {
+	Start int
+	Stop  *int
+}
+
+type JSONArrTrimArgs struct {
+	Start int
+	Stop  *int
+}
+
+type JSONCmd struct {
+	baseCmd[string]
+	expanded []any // expanded will be used at JSONCmd.Expanded
+	typ      jsonCmdTyp
+}
+
+type jsonCmdTyp int
+
+const (
+	TYP_STRING jsonCmdTyp = iota
+	TYP_ARRAY
+)
+
+// https://github.com/redis/go-redis/blob/v9.3.0/json.go#L86
+func (cmd *JSONCmd) Val() string {
+	return cmd.val
+}
+
+// https://github.com/redis/go-redis/blob/v9.3.0/json.go#L105
+func (cmd *JSONCmd) Expanded() (any, error) {
+	if cmd.typ == TYP_STRING {
+		return cmd.Val(), nil
+	}
+	// TYP_ARRAY
+	return cmd.expanded, nil
+}
+
+func (cmd *JSONCmd) Result() (string, error) {
+	return cmd.Val(), nil
+}
+
+func newJSONCmd(res rueidis.RedisResult) *JSONCmd {
+	cmd := &JSONCmd{}
+	msg, err := res.ToMessage()
+	if err != nil {
+		if err == rueidis.Nil {
+			cmd.SetVal("")
+			return cmd
+		}
+		cmd.SetErr(err)
+		return cmd
+	}
+	switch {
+	// JSON.GET
+	case msg.IsString():
+		cmd.typ = TYP_STRING
+		str, err := res.ToString()
+		if err != nil {
+			cmd.SetErr(err)
+			return cmd
+		}
+		cmd.SetVal(str)
+	// JSON.NUMINCRBY
+	case msg.IsArray():
+		// we set marshaled string to cmd.val
+		// which will be used at cmd.Val()
+		// and also stored parsed result to cmd.expanded,
+		// which will be used at cmd.Expanded()
+		cmd.typ = TYP_ARRAY
+		arr, err := res.ToArray()
+		if err != nil {
+			cmd.SetErr(err)
+			return cmd
+		}
+		expanded := make([]any, len(arr))
+		for i, e := range arr {
+			anyE, err := e.ToAny()
+			if err != nil {
+				cmd.SetErr(err)
+				return cmd
+			}
+			expanded[i] = anyE
+		}
+		cmd.expanded = expanded
+		val, err := json.Marshal(cmd.expanded)
+		if err != nil {
+			cmd.SetErr(err)
+			return cmd
+		}
+		cmd.SetVal(string(val))
+	default:
+		panic("invalid type, expect array or string")
+	}
+	return cmd
+}
+
+type JSONGetArgs struct {
+	Indent  string
+	Newline string
+	Space   string
+}
+
+type IntPointerSliceCmd struct {
+	baseCmd[[]*int64]
+}
+
+// newIntPointerSliceCmd initialises an IntPointerSliceCmd
+func newIntPointerSliceCmd(res rueidis.RedisResult) *IntPointerSliceCmd {
+	cmd := &IntPointerSliceCmd{}
+	arr, err := res.ToArray()
+	if err != nil {
+		cmd.SetErr(err)
+		return cmd
+	}
+	intPtrSlice := make([]*int64, len(arr))
+	for i, e := range arr {
+		if e.IsNil() {
+			intPtrSlice[i] = nil
```

Comment:
This also seems unnecessary.

Reply:
Do you mean `intPtrSlice[i] = nil` is unnecessary because element of intPtrSlice has default value `nil` ?

Reply:
Yes.